### ISSUE TITLE
[Obsidian] Vault not found (but location is correct)

### ISSUE THREAD

**Issue by sammcj** [raycast/extensions#11741]
### Extension

https://www.raycast.com/KevinBatdorf/obsidian
https://github.com/raycast/extensions/tree/main/extensions/obsidian

### Description

 The official Raycast extension for Obsidian fails to find my Vault.

I have tried:

- A vault located in the standard iCloud location (`/Users/samm/Library/Mobile Documents/iCloud~md~obsidian/Documents/Main`)
- A vault located in iCloud drive (`/Users/samm/Library/Mobile Documents/com~apple~CloudDocs/Obsidian/Main`)
- Creating a new Vault
- Quoting and unquoting the path
- Adding/removing \ before the space in the path
- Updating all plugins, Raycast, Obsidian

![SCR-20240411-hlwb](https://github.com/raycast/extensions/assets/862951/01ef1bfa-02b3-46b7-ada1-778ca04005bb)

![SCR-20240411-hlxy](https://github.com/raycast/extensions/assets/862951/fcfe9c05-d8cc-4c08-8d93-34e4ce5cc291)

![SCR-20240411-hlzm](https://github.com/raycast/extensions/assets/862951/4da37891-4b59-4e2a-a70f-c2bf4c27e4e2)

![SCR-20240411-hovt](https://github.com/raycast/extensions/assets/862951/f157a974-24ea-4480-8769-a54ce89f61cf)

- Adding the parent path (containing multiple vaults) - this results in a different error

![SCR-20240411-hofz](https://github.com/raycast/extensions/assets/862951/d2a9db6d-e5fb-4024-b565-df272b9fedce)

### Steps To Reproduce

1. Create a new Obsidian vault in the default location (`/Users/Username/Library/Mobile Documents/iCloud~md~obsidian/Documents/VaultName`
2. Add the path to the extensions settings
3. Try to search for a note in the vault

### Current Behaviour

No vaults found

### Expected Behaviour

Vault(s) found

### Logs

```
Error: Worker terminated due to reaching memory limit: JS heap out of memory

NodeError:errors:406:5
[kOnExit]:worker:313:26
Worker.<computed>.onexit:worker:229:20
```

**Comment by sammcj** [comment]
@raycastbot it's linked in the issue?

**Comment by sammcj** [comment]
oo maybe similar to https://github.com/raycast/extensions/issues/11445

**Comment by KevinBatdorf** [comment]
Thanks for reporting. For debugging purposes, how many notes do you have?

I believe the issue is that the extension loads the entire vault into memory to perform searches, and if you have a lot of notes, or possibly enough large notes, it crashes. Do you happen to know the file size of your largest note?

The bottom line is that I need to find the right time to fix this as I think it's going to be a bit complicated, and take time to try out a few different ideas that might lead to nothing. (Note to readers: if someone else wants to take this on I'm happy to guide, code review, etc - beginners are welcome too). 

I'm hoping there's a way to just execute a shell command to grep the contents, but I'll have to read other extensions to see how it's being done there. This might not solve it.

Alternatively, it might require creating a database that syncs the notes content locally. I'd like to avoid this duplication but it could lead to a better search experience than native obsidian. 

@pernielsentikaer you mentioned previously raycasts new search. I searched the docs for a new feature but didn't find anything. Do you have any suggestions? How would the team recommend searching an entire large directory and all its contents?

**Comment by sammcj** [comment]
Howdy Kevin,

> the extension loads the entire vault into memory

Oh my...

I don't have that big of a vault compared to others I know, but it certainly seems far too big to have loaded into memory just to open/search notes.

<img width="322" alt="image" src="https://github.com/raycast/extensions/assets/862951/baf5590f-4699-4400-aef8-d6473f0c19c3">

Currently at 2385 but will only keep growing.

> The bottom line is that I need to find the right time to fix this as I think it's going to be a bit complicated

Totally understood, no expectations, open source hugs etc etc etc

I'll have a look at the Obsidian API, while I'm no web-dev I've been known to hack up some JS/TS now and then. If I do come up with something I'll let you know.

**Comment by sammcj** [comment]
Actually, this looks interesting: https://github.com/coddingtonbear/obsidian-local-rest-api?tab=readme-ov-file

**Comment by KevinBatdorf** [comment]
It appears they search through every file one by one.

https://github.com/coddingtonbear/obsidian-local-rest-api/blob/216bffcc4739e9428c71396c1b06e9dcd7d093ea/src/requestHandler.ts#L787

That's one idea I had but it could potentially be slow as well. I didn't realize there was an obsidian sdk actually so maybe the cached file and search builder they have there makes it faster.

Could you try that rest library out and tell me if it's fast on your vault?

Implimenting this would require creating custom search in raycast which is fine and I've done before elsewhere, but I just worry if it will be slowing things down for the 99% too. We could make it opt-in though as well. Better than just crashing!

**Comment by sammcj** [comment]
I just tried it out with some standalone code and it's _very_ fast, less than a second to perform a search across my whole database.

```typescript
import axios from 'axios';
import https from 'https';

const API_URL = 'https://127.0.0.1:27124'; // Replace with the appropriate URL and port
const API_KEY = 'REDACTED'; // Replace with your actual API key

// Create an instance of the https agent
const httpsAgent = new https.Agent({
  rejectUnauthorized: false, // Allow self-signed certificates
});

// Configure axios with the base URL, authorization header, and httpsAgent
const api = axios.create({
  baseURL: API_URL,
  headers: {
    'Accept': '*/*',
    'Authorization': `Bearer ${API_KEY}`,
    'Content-Type': '*;/*',
    // 'Content-Type': ' text/markdown',
  },
  httpsAgent: httpsAgent, // Use the httpsAgent to allow self-signed certificates
});

// define an interface for a document
interface Document {
  filename: string;
  content: string;
  score: number;
  matches: unknown[];
}

// Function to discover configured vaults
async function getFileContent(filename: string): Promise<string> {
  try {
    const response = await api.get(`/vault/${filename}`);
    return response.data;
  } catch (error) {
    console.error(`Error getting file content: ${filename}`);
    throw error;
  }
}

async function createOrUpdateFile(filename: string, content: string): Promise<void> {
  try {
    await api.put(`/vault/${filename}`, content, {
      headers: {
        'Content-Type': 'text/markdown',
      },
    });
    console.log(`File "${filename}" created or updated successfully.`);
  } catch (error) {
    console.error(`Error creating or updating file: ${filename} - error: ${error}`);
    throw error;
  }
}

async function searchDocuments(query: string): Promise<Document[]> {
  try {
    const response = await api.post('/search/simple/', null, {
      params: {
        query: query,
      },
    });

    // create a map of the filenames, scores, and matches
    const documents: Document[] = response.data;
    const results = documents.map((document) => ({
      filename: document.filename,
      score: document.score,
      matches: document.matches,
    }) as Document);

    return results;


  } catch (error) {
    console.error(`Error searching documents: ${query} - error: ${error}`);
    throw error;
  }
}

// Example usage
(async () => {
  try {

    // Create or update a file
    await createOrUpdateFile('testfile.md', '# New File\n\nThis is the content of the new file.');

    // Get the content of a file
    const fileContent = await getFileContent('testfile.md');
    console.log('File content:\n\n', fileContent);

    // Search for documents
    const searchResults = await searchDocuments('query');
    console.log('Search results:', searchResults);
  } catch (error) {
    console.error('An error occurred:', error);
  }
})();
```

```
  {
    filename: 'Bear/You Are Not Google – Bradfield.md',
    score: -37.5022,
    matches: [ [Object], [Object], [Object] ]
  },
  {
    filename: 'Bear/Scientists rise up against statistical significance.md',
    score: -36.9055,
    matches: [ [Object], [Object] ]
  },
  {
    filename: 'Bear/Five Tips For a Healthier Postgres Database in the New Year.md',
    score: -36.64960000000001,
    matches: [ [Object], [Object], [Object], [Object] ]
  },
  {
    filename: 'Bear/Facebook’s war on free will  Technology  The Guardian.md',
    score: -32.612500000000004,
    matches: [ [Object] ]
  },
  ... 131 more items
]
node api.js  0.09s user 0.01s system 50% cpu 0.196 total
```

**Comment by KevinBatdorf** [comment]
Ok then a Raycast custom search implimentation (which might already be implemented actually) using the official sdk search wrapper, and iterating over every file seems likely to be the best approach.

Thanks for investigating that. I'll take a closer look at it. I'll try for this weekend and will follow up.

**Comment by KevinBatdorf** [comment]
If you also wanted to take a look, I'll describe the important files (I didn't initially write this code so I may miss something)

The [NoteLoader](https://github.com/raycast/extensions/blob/main/extensions/obsidian/src/utils/data/loader.tsx) class loads in all the notes and media items for a vault, and is consumed by the [cache mechanism](https://github.com/raycast/extensions/blob/main/extensions/obsidian/src/utils/data/cache.tsx). Thayt's then used by the [useNotes](https://github.com/raycast/extensions/blob/main/extensions/obsidian/src/utils/hooks.tsx#L6) hook, which is loading the cache (so all the notes) into memory.

`useNotes` is used throughout the extension. This alone could probably be optimized though as (I could be wrong here) I believe it's loading a new copy of the vault every time the hook is invoked in a component (which it may not be invoked that often per command lifecycle anyway).

There needs to be a list of notes though that show up _before_ a search takes place so the user can see recent notes, navigate up/down, etc, but we could also have some sort of lazy loading mechanism that fetches notes on demand as you near the end of the list (so 50 notes on start, then when the users scrolls to note 40 we fetch the next set of 50, etc etc)

For searching, there already is a custom search implimentation [here](https://github.com/raycast/extensions/blob/main/extensions/obsidian/src/components/NoteList/NoteList.tsx#L37), but it searches the cache of notes mentioned above. An interation on this would involve removing this and adding a search mechanism more simlar to what we just talked about.

So in summary,

1. We could try to confirm whether the valut cache is being loaded into memory multiple times and optimize that, and that might actually be a good enough solution for now. It would also be fairly easy to fix as it would just involve editing the useNotes hook. It might not help if someone has 1 million notes though...

After that, we need to explore
1. Removing rhe cache mechanism
2. Impliment lazy loading of notes as the user scrolls down
3. Impliment a search using the sdk's search helpers and iterating over every file one by one

**Comment by ykhalfalli** [comment]
Running into this issue myself and also totally new to GH so will this comment reopen the issue? Thanks in advance and let me know what you'd need from me to help
